### ISSUE TITLE
Parallelization when `lse_mode=False`

### ISSUE THREAD

**Issue by gjhuizing** [ott-jax/ott#371]
Hello!

In POT, when using the vector product Sinkhorn iterations, you can compute several distances efficiently by replacing the vectors `u` and `v` with matrices. When using log iterations, they just replace this with a for loop: https://pythonot.github.io/_modules/ot/bregman.html#sinkhorn2

As shown in the Colab below, a simple vmap on the second marginal does not achieve the same efficiency in OTT. Is there a smarter way to do this? 

I guess adapting the code would be quite heavy. It seems JAX will support custom batching rules at some point, so maybe that would be a more elegant way ?

https://colab.research.google.com/drive/1x0v2dvYxxHHs8DgkgCKz1l63gX9QBsJg?usp=sharing

Best,

GJ

**Comment by marcocuturi** [comment]
Hi @gjhuizing 

Thanks for the nice example.

A few first comments 
- you used `max_norm` to normalise, but you should have used`max_cost`. Checking this both versions are now equivalent (this one was noticed by @michalk8 !)
- OTT's sinkhorn uses 1-norm by default to track convergence, POT uses 2-norm, so the thresholds were not comparable. You can specify the `norm_error` parameter to be 2 for both to be comparable.
- Also, your runtime for POT did not account for the cost of pre-materialising the cost matrix.
- In the setting you were considering (`epsilon=1e-1` and `threshold=1e-2`) and after taking into account modifications above, both algorithms only needed 1 Sinkhorn iteration. In that setting POT is lighter and takes slightly less time to run. Switching to respectively 1e-2 and 1e-4 you will see in my colab  that when we are running more Sinkhorn iterations OTT becomes faster.

https://colab.research.google.com/drive/1iur7FHeKGJEqccArPeJLwfvUffV3pP0r?usp=sharing

Now, w.r.t. to parallelisation, it is true that we have not implement what I was advocating in [my paper](https://papers.nips.cc/paper_files/paper/2013/file/af21d0c97db2e27e13572cbf59eb343d-Paper.pdf) (Alg. 1 more specifically), because it does not work in `lse_mode`. 

In principle, using kernel mode should allow for vmapping, but my guess is that this is not working efficiently because we are not pooling those kernel multiplies, and each Sinkhorn has its own termination criterion (this might slow other executions, you can see in the nb that execution might a bit slower when repeating the same computation). Because of memory I also suspect that computations are queued. 

In older implementations (e.g. https://github.com/google-research/ott/releases/tag/0.1.20) we were accepting `[batch, n]` or `[batch, m]` weights but if I recall correctly we dropped this.

it might make sense to reinstate this in a way or another... and not that challenging since our `Geometry.apply_kernel` operators are parallel by nature. Things might breakdown though for other stuff (acceleration in particular), so it might be better to use a different API (like the `segment_sinkhorn...` that we have.

**Comment by gjhuizing** [comment]
Thanks a lot @marcocuturi for the detailed answer!

- What exactly does `max_norm` do? Looking at the code, it seems it takes the max of the norms of the points in the point cloud? This could be good to add in the docstring.
- I was wondering if the stopping criterion was the same, great to know you can change it! Ideally, this should be done in the [OTT vs POT](https://ott-jax.readthedocs.io/en/latest/tutorials/notebooks/OTT_%26_POT.html) tutorial and in the upcoming benchopt don't you think?
- Same for the threshold, I used the same as the tutorial. But maybe it's not a big deal since epsilon is smaller.

Indeed a separate API like `segment_sinkhorn` could be nice and would not break anything else. Arguably it's an exotic use case but in [this](https://academic.oup.com/bioinformatics/article/38/8/2169/6528312) and [this](https://arxiv.org/abs/2102.06278) I was computing pairwise OT distance matrices, which currently would be too slow with OTT.

[This ott-jax tutorial](https://ott-jax.readthedocs.io/en/latest/tutorials/notebooks/One_Sinkhorn.html) also computes pairwise distances.

If I were to play around with this idea, where should I start? A subclass of the `Sinkhorn` solver and a small wrapper in `tools`?

**Comment by marcocuturi** [comment]
Thanks!
- `max_norm` does what you described. It's used for geometries that are no passed explicitly, or that cannot be materialized efficiently. Imagine for instance you are using `PNormP` with `p=1.5`, the `batch` option to never store the cost matrix. In that case materializing the $nm$ matrix to computing its mean or max is too expensive. Computing $n+m$ norms is easier. This is what we do. It's as if you were normalizing by diameter.
- you are right, I made a mistake in the tutorial... I think that, at that time, I had set (temporarily) the default to 2-norm. I think 1-norm makes more sense. I will refresh that tutorial
- I am still thinking about a better API. Maybe the simplest is to vectorize explicitly code by reintroducing this batch dimension. I think from easy to hard we would have
  - sinkhorn steps in `lse_mode=False` (fairly trivial because all `apply_kernel` methods are batched by default)
  - computation of errors (same), although this introduces a difference in stopping criterion (maybe a max of deviations, or average as done in POT).
  - implicit differentiation (harder, requires applying  $b$ transport applications in each sparse tensor, here the `geom` is the same, which can save something... I imagine this would lead to a big $b^2n^2$ block diagonal linear operator, inverting $bn$ vectors simultaneously)

**Comment by gjhuizing** [comment]
If you think it's worth it I can give it a try! (at least the first two points)

**Comment by marcocuturi** [comment]
Thanks for proposing this! I am cleaning up the `POT vs OTT` nb and also want to run it with their jax backend.

Before we embark , I think there's an easy fix. If you set `min_iterations=xxx, max_iterations=xxx` in the parameters of your `Sinkhorn` solver, `ott-jax` switches to a `scan` loop and then XLA should be able to compile this more efficiently and return good parallelisation. The downside is that this is more `geom_loss`-like, no proper control on convergence.

right now the convergence flagged would be bugged though (because of #375).

you can give this a try, or I'll do it myself when I have a minute.

**Comment by gjhuizing** [comment]
Ah that sounds great, I'll try it and see if it speeds things up :)

**Comment by marcocuturi** [comment]
thanks again @gjhuizing for bringing this to our attention!!!

**Comment by gjhuizing** [comment]
Amazing, thank you! Great that there was such an easy fix :)
